Added prime testing algo of Euler

rohitx007 opened the pull request:
This gives Euler Pseudoprime numbers now
With help of this, I will extend it to include Solovay-Strassen primality test

rohitx007 committed:
Added quadraticresidue.py

rohitx007 committed:
Added_Euler_Psudoprime_method.py

Will add Solvay Strassen primality test using Euler Pseudoprime test

rohitx007 committed:
Update modular.py

normalhuman commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,19 @@ def mr(n, bases):
 
 
+def isprimeE(n,b):
+	"""isprimeE(n) - Test whether n is prime or an Euler pseudoprime to base b."""
```
docstrings usually do not repeat the name of the function. """Test whether n is prime or an Euler pseudoprime to base b.""" would be enough.

rohitx007 replied:
Please do comment unnecessarily on such trivial things if there is any conceptual/coding mistake then only comment

normalhuman commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,19 @@ def mr(n, bases):
     return True
 
 
+def isprimeE(n,b):
```
Here and below, there should be a space after each comma

rohitx007 replied:
This is readable and follows basic standards of code

normalhuman commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,19 @@ def mr(n, bases):
 
+def isprimeE(n,b):
+	"""isprimeE(n) - Test whether n is prime or an Euler pseudoprime to base b."""
+	if (not mr(n,b)): return False
```
Here and below, there should be a new line after  colon  (PEP8, you can check it online with http://pep8online.com/)

The parentheses around if conditions are not needed.

rohitx007 replied:
Parenthesis are necessary for readability when writing the result of the condition and the case in one line

rohitx007 replied:
I will change and add a new line after the colon according to PEP8.
thanks...

rohitx007 committed:
Added space after comma

added spaces and increased PEP8 readability according to previous comments

smichr commented on sympy/ntheory/primetest.py:
```diff
@@ -110,17 +110,22 @@ def mr(n, bases):
-	"""isprimeE(n) - Test whether n is prime or an Euler pseudoprime to base b."""
-	if (not mr(n,b)): return False
+def isprimeE(n, b):
+    """isprimeE(n ,b) - Test whether n is prime or an Euler pseudoprime to base b."""
```
@normalhuman gave you feedback that is consistent with our desire for code in the SymPy codebase. Please see [theis PEP](https://www.python.org/dev/peps/pep-0257/) if you have questions about docstrings.  While the code is what "get's it done" the docstring is the welcome mat for the public. Your docstring is fine past the first space after the dash.

rohitx007 replied:
done.
thanks for review..

smichr commented on sympy/ntheory/primetest.py:
```diff
@@ -110,17 +110,22 @@ def mr(n, bases):
-	if (not mr(n,b)): return False
+def isprimeE(n, b):
+    """isprimeE(n ,b) - Test whether n is prime or an Euler pseudoprime to base b."""
+	if (not mr(n, b)): return False
```
Please make the return on its own line here and below. This makes code coverage easier to see/verify.

rohitx007 replied:
Sure Sir !
Done

smichr commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,24 @@ def mr(n, bases):
+	if (not mr(n, b)): return False
+    
+	r = n-1
+	while (r % 2 == 0): r //= 2
```
Make sure you are working with ints and then this can be `r >>= trailing(r)`

rohitx007 replied:
Yeah, I have changed it now,
thanks for the review!!

smichr commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,24 @@ def mr(n, bases):
+	c = pow(b, r, n)
+	if (c == 1): return True
+	while (1):
+		if (c == 1): return False
```
This test for `c == 1` can appear after recomputing the c since on the first pass we know it can't be true.

rohitx007 replied:
nice observation sir!
will fix it for sure

smichr commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,24 @@ def mr(n, bases):
 
+def isprimeE(n, b):
+    """isprimeE(n ,b) - Test whether n is prime or an Euler pseudoprime to base b."""
+	if (not mr(n, b)): return False
```
`mr` expects a list of bases so this should be passed as `mr(n, [b])`.

rohitx007 replied:
Sorry for this silly typo.
Its fixed now thanks!

rohitx007 committed:
updated changes according to feedback 

fixed some typos and list errors

smichr commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,31 @@ def mr(n, bases):
+	if (not mr(n, [b])):
+        return False
+    
+    r = int(r)
```
indentation problem and r has not yet been defined. See `mr` for a way to check that n and b are integers by using `as_int`.

rohitx007 replied:
Yes sir,
Thanks or suggestion Its updated now.

smichr commented:
Looks like the other changes didn't get committed. I'll kill the Travis run.

asmeurer commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,31 @@ def mr(n, bases):
 
+def isprimeE(n, b):
+    """Test whether n is prime or an Euler pseudoprime to base b."""
+	if (not mr(n, [b])):
```
Don't use parentheses after `if` or `while`.

rohitx007 replied:
Okay
I have removed parenthesis now

asmeurer commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,31 @@ def mr(n, bases):
+	if (c == 1):
+        return True
+    
+	while (1):
```
Use `while True`.

rohitx007 replied:
Yeah technically while True is more correct,
Thanks for the suggestion.

asmeurer commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,31 @@ def mr(n, bases):
 
 
+def isprimeE(n, b):
+    """Test whether n is prime or an Euler pseudoprime to base b."""
```
It could be useful to put the definition on a second line

rohitx007 replied:
Done!

asmeurer commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,31 @@ def mr(n, bases):
     return True
 
 
+def isprimeE(n, b):
```
isprime_euler or is_euler_pseudoprime would be a better name.

rohitx007 replied:
Yeah it sounds more relating
I changed it
Thanks for the suggestion.

rohitx007 committed:
Typo error.py

minor typo error fixed

rohitx007 committed:
Update primetest.py

smichr commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,34 @@ def mr(n, bases):
+    
+    
+	r = n-1
+    r = as_int(r)
```
- [ ] indentation is off; I will kill the Travis run.
- [ ] Since you use `n` later, make n an int via `n = as_int(n)`; then calculations done with it will still be int.
- [ ] Use `trailing` as I said before.
- [ ] Don't use `(` after `if` in the remaining tests.
- [ ] Move the `if c == 1: return False` to *after* the recalculation of `c`.

rohitx007 replied:
Sir ,
Now  I have done following changes to code 
1] I have put the docstring inside the function now according to suggestion by smichr
2] indented properly as previous error occurred due to mix of tabs and space which wasn't visible in github sorry for trouble
3] made n=as_int(n)
4]Used trailing from sympy factor to calculate ending zeros of r instead of while loop
5]removed bracket after if following PEP standards.
6]Moved the if c == 1: return False to after the recalculation of c.
Thanks for help and suggestions.

rohitx007 committed:
Changed all the errors according to feedback

1] I have put the docstring inside the function now according to suggetion by smichr
2] indented properly as previous error occured due to mix of tabs and space which wasnt visible in github sorry for trouble
3] made n=as_int(n)
4]Used trailing from sympy factor to claculate ending zeros of r instead of while loop
5]removed bracket after if following PEP standards.
6]Moved the if c == 1: return False to after the recalculation of c.

smichr commented on sympy/ntheory/primetest.py:
```diff
@@ -110,6 +110,32 @@ def mr(n, bases):
+    c = pow(b, r, n)
+    if c == 1:
+        return True
+    
```
It looks like you are not that familiar with Python requirements or else are using an editor that is inserting indentation for you. The quality tests will not pass if there is trailing whitespace on a line (as there is on line 128).

I recommend the following code for what you have written:

```
def is_euler_pseudoprime(n, b):
    """Return True if n is prime or an Euler pseudoprime to base b, else False."""
    from sympy.ntheory.factor_ import trailing

    if not mr(n, [b]):
        return False

    n = as_int(n)
    r = n - 1
    c = pow(b, r >> trailing(r), n)
    if c == 1:
        return True

    while True:
        if c == n - 1:
            return True
        c = pow(c, 2, n)
        if c == 1:
            return False

```

smichr replied:
Also, since there are 4 return statements, you should add 4 tests: one pair of n,b that return from each of those `return` lines. The tests should go in `test_primetest.py` in the tests folder of ntheory.

But...before you do more. What will the ` Solovay-Strassen` test do that is an improvement over what is already in SymPy? I ask because my quick read of the wiki on the method says that the Miller-Rabin test generally is preferred: "It has been largely superseded by the Baillie-PSW primality test and the Miller–Rabin primality test, but has great historical importance" [ref](https://en.wikipedia.org/wiki/Solovay%E2%80%93Strassen_primality_test).

rohitx007 replied:
Sir,
I agree with your point on M-R being better than Solovay-Strassen but,
M-R is a probabilistic test which is based on the assumption of the Extended Riemann Hypothesis which no one has yet proved to be correct.
Also, all prime tests are probabilistic in sympy.My main idea is to add a deterministic prime test like AKS and/or APR-CL with runtime less than miller rabin test !!!!
https://en.wikipedia.org/wiki/Adleman%E2%80%93Pomerance%E2%80%93Rumely_primality_test

For these I need Euler pseudoprimes

rohitx007 committed:
Update primetest.py

smichr commented:
Look slike too many blannk lines at the end of series.py

Note: you can run the quality tests locally before pushing changes to make sure that at least *that* error won't waste a test cycle.

```
AssertionError: File ends with more than 1 newline: /home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg/sympy/series/series.py, line 13
```
